enhance Threads.Events with ordering and (auto)reset

vtjnash opened the pull request:
Previously, this was using memory-order relaxed, though that is unlikely
to be useful. Instead guarantee it will have memory-order
acquire-release (similar to a Lock).

vtjnash committed:
enhance Threads.Events with ordering and (auto)reset

Previously, this was using memory-order relaxed, though that is unlikely
to be useful. Instead guarantee it will have memory-order
acquire-release (similar to a Lock).

JeffBezanson commented on base/lock.jl:
```diff
@@ -342,28 +346,44 @@ end
-        while !e.set
+        if !e.set
             wait(e.notify)
+            @atomic e.set # full barrier also
```
What does this do?

vtjnash replied:
ah, this is leftover now from a time when I considered making the ordering guarantee even stronger

JeffBezanson commented on base/lock.jl:
```diff
@@ -342,28 +346,44 @@ end
-tasks will no longer block when waiting for it.
+tasks will no longer block when waiting for it, until `reset` is called.
+
+If `autoreset` is true, at most one thread will be released from `wait` for
```
```suggestion
If `autoreset` is true, at most one task will be released from `wait` for
```
?

JeffBezanson commented on base/lock.jl:
```diff
@@ -342,28 +346,44 @@ end
             wait(e.notify)
+            @atomic e.set # full barrier also
+        elseif e.autoreset
+            @atomic e.set = false
```
I think this needs to be an atomic swap, and then possibly go around the `while` loop again, in case another thread reaches `wait` just before this line?

JeffBezanson replied:
Aha, or I guess in that case `e.set` won't have been set, so it's ok?

vtjnash replied:
There's a lock here, so that situation can and must never occur

vtjnash replied:
ooh, I see it might get cleared outside this loop, okay

tkf commented:
Hmm... Is autoreset-`Event` used in other concurrent platforms? It seems to weaken level-trigger-ness of `Event`? For example, IIUC, if we have two `notify` calls and two waiters, one of the waiters may not be notified depending on the timing.

We already have `Channel{Nothing}` as a synchronization API that a "notify" exactly matches a "wait." There's also `Condition` as a lower-level API. So, I can't think of a pattern where autoreset-`Event` fits nicely at the moment. Maybe some motivating example in the documentation helps?

tkf commented:
I wonder if Tokio's watch channel https://docs.rs/tokio/0.1.22/tokio/sync/watch/index.html is somewhat close to this if we generalize autoreset-`Event` to "set-many promise?"

vtjnash commented:
Yes, it replaces the level-triggeredness with edge-triggeredness. I have found this to be useful in some situations, such as our Timer and AsyncCondition implementations. It is an infrequent paradigm perhaps, but one that allows expressing lock-free edge-triggered condition variables without race conditions even for multi-producer multi-consumer async situations. It is used in places where that matters (usually kernel APIs, where locks cannot be held) such as kqueue, epoll, and [Event](https://docs.microsoft.com/en-us/windows/win32/sync/event-objects) objects. It is perhaps also similar to a unix `signal`, in that updates may be coalesced (unlike Channel or Tokio.watch).

vtjnash committed:
enhance AsyncCondition with ordering promise

Previously, this was using memory-order relaxed, though that is unlikely
to be useful. Instead guarantee it will have memory-order
acquire-release (similar to an Event).

vtjnash commented:
Made some more changes to be clear about where I think the ordering is guaranteed by, and fixed AsyncEvent too.

tkf commented on base/lock.jl:
```diff
@@ -342,40 +346,62 @@ end
 
 
 """
-    Event()
+    Event([autoreset=false])
 
 Create a level-triggered event source. Tasks that call [`wait`](@ref) on an
```
I wonder if we can use a different name though? It's conceptually much cleaner if we can say `Condtion` is edge-triggered and `Event` is edge-triggered. We can still share most of the implementation by using `::Union{Event,AutoreSetThingy}`. For example, would it be crazy to call this a `Signal`? Or, if we go with `Event([autoreset=false])`, I think we need to at least reword the dosctring "Create a level-triggered event source."

vtjnash replied:
Being level-triggered doesn't preclude being resettable, though I agree the historic choices of names was unfortunate, I don't think we should attempt to deviate from them now.

tkf replied:
My point was the opposite. I was not suggesting moving away from the historical naming.

Rather, one of my suggestions was to introduce a new type for `Event(true)` and keep the old semantics of `Event()` as-is.

vtjnash replied:
This doesn't change the semantics of `Event()`, but extends it with a useful feature (the ability to reset it manually/automatically) common to other uses of the name (in kevent, the name of this flag is EV_CLEAR, in Event, the name of this flag is bManualReset)

tkf replied:
Hmm..., I think it changes how much you can infer locally about the correctness of the code. IIUC, with `autoreset = false`, I can conclude that

```julia
ev::Event
ref::Ref
@sync for i in 1:2
    @spawn begin
        wait(ev)
        ref[] = i
    end
end
```

has a data race. With `autoreset = true`, it may not have a data race (depending on the use of `notify`). This is obviously not the best example. But, I think it at least shows that this PR does change the semantics of `Event`.

I can't think of any reasons for avoiding different types to represent `autoreset`-or-not other than coming up a good name is hard (which *can* also be reflected to the discoverability of the API; however, I'd assume this is negligible as long as the new API is linked from `Event` docstring). It's hard to come up with a decent name but that's a one time cost. Readability of the code has an effect to the cost proportional to the time and the number of Julia programmers using it.

JeffBezanson replied:
This feature seems OK to me. `Event` is still level-triggered; the only difference is that the level might not last forever. It should be clear from context which events might change status, e.g. "data available" vs. " task has exited".

tkf replied:
Is there a downside in providing the information on autoreset-or-not as a type in the code (e.g., `Event` vs `AutoresetEvent`)? Concurrent programming is especially tricky and so more information (which is often locally obvious) seems to be strictly better to me.

That said, I don't have enough experience in using autoreset-`Event` and so I don't know if the upside I'm arguing is negligible. Also, users can always create a wrapper type `AutoresetEvent` that ensures `Event(true)` is called in the constructor if they really need it (but this makes reading others' code harder).

tkf reviewed (commented):
> It is used in places where that matters (usually kernel APIs, where locks cannot be held) such as kqueue, epoll, and [Event](https://docs.microsoft.com/en-us/windows/win32/sync/event-objects) objects. It is perhaps also similar to a unix `signal`, in that updates may be coalesced

Thanks for the reference! I think get the similarity to unix signal. I remember also seeing that @vchuravy and @JeffBezanson were discussing that `Event` couldn't be used in Distributed.jl as it's not set-many https://github.com/JuliaLang/julia/pull/38405. So I guess I think I get a very rough idea and usefulness.

vtjnash committed:
enhance jl_critical_error with more state reset

We want to fail task-switches here for better reasons than an excess of locks being held during atexit callbacks.

tkf reviewed (changes requested):
FWIW, I don't think my concern https://github.com/JuliaLang/julia/pull/43095#discussion_r750653969 is resolved. I believe this change makes it difficult to explain what `Event` is. It also makes reading code involving `Event` strictly harder since you can't tell which mode you are using from the type and these modes have very different synchronization semantics.

tkf reviewed (approved):
I don't agree with the approach that `autoreset` is a run-time flag. But I'm not the one who needs this ATM and so I don't feel like I should be blocking this.
